### ISSUE TITLE
[EKS/Fargate] [request]: Transparent EKS fargate management

### ISSUE THREAD

**Issue by khacminh** [aws/containers-roadmap#807]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
Transparent management for EKS Fargate

**Which service(s) is this request for?**

EKS Fargate

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

### Limit of Fargate log

Fargate on EKS is a great idea and I really enjoyed the first time with it. However, everything getting bad when I upgraded my EKS cluster to 1.15. Right after the control plane updated to version 1.15, the new fargate pod cannot be scheduled on Fargate any more and they are stuck as pending. I spent days troubleshooting with no luck. Then, I realized that I was working with it as a blind man. The only thing I found about this problem on AWS documentation is:

> Pods which do not match a Fargate profile may be stuck as Pending. If a matching Fargate profile exists, you can delete pending pods that you have created to reschedule them onto Fargate

In my case, no useful log could be found. EKS Fargate adds the new scheduler named **fargate-scheduler** but I couldn't find the log of this scheduler on Cloudwatch. Even now, when I can be sure about the fargate bug, I can't dig it deeper.

### No resource visualize

 On EKS 1.14, I can schedule my pods on fargate. However, nothing is visualized on the console page. I think the metrics like Lambda would be really helpful.

**Comment by nicholasgcoles** [comment]
@khacminh what was the result when you described your pods? `kubectl describe pod ...`

**Comment by khacminh** [comment]
Hi @nicholasgcoles,

Here is the result when run `kubectl describe pod ...`

```yaml
# kubectl describe pod test-fargate-app-6b4fb446c6-rw2pj

Name:               test-fargate-app-6b4fb446c6-rw2pj
Namespace:          mynamespace
Priority:           2000001000
PriorityClassName:  system-node-critical
Node:               <none>
Labels:             app=test-fargate-app
                    deploy-env=fargate
                    eks.amazonaws.com/fargate-profile=my-fargate-profile
                    pod-template-hash=6b4fb446c6
Annotations:        kubernetes.io/psp: eks.privileged
Status:             Pending
IP:
Controlled By:      ReplicaSet/test-fargate-app-6b4fb446c6
NominatedNodeName:  c62032c57f-3bf20f49a6d644568d59216432b8f221
Containers:
  test-fargate-app:
    Image:      xxxxxxxxxxxx.dkr.ecr.us-east-2.amazonaws.com/simple-app:0.1.0
    Port:       80/TCP
    Host Port:  0/TCP
    Args:
      /bin/sh
      -c
      echo hello;sleep 3600
    Limits:
      cpu:     250m
      memory:  512Mi
    Requests:
      cpu:        250m
      memory:     512Mi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-bpxsq (ro)
Volumes:
  default-token-bpxsq:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-bpxsq
    Optional:    false
QoS Class:       Guaranteed
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

And this one is the result of `kubectl get pod ... -o yaml`

```yaml
----
# kubectl get pod test-fargate-app-6b4fb446c6-rw2pj -o yaml

apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/psp: eks.privileged
  creationTimestamp: "2020-03-27T01:25:05Z"
  generateName: test-fargate-app-6b4fb446c6-
  labels:
    app: test-fargate-app
    deploy-env: fargate
    eks.amazonaws.com/fargate-profile: my-fargate-profile
    pod-template-hash: 6b4fb446c6
  name: test-fargate-app-6b4fb446c6-rw2pj
  namespace: mynamespace
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: test-fargate-app-6b4fb446c6
    uid: a074c6a5-9753-4bf0-b169-545c58863aae
  resourceVersion: "68357633"
  selfLink: /api/v1/namespaces/mynamespace/pods/test-fargate-app-6b4fb446c6-rw2pj
  uid: 92d5485e-da18-4581-a680-c56c74b148ab
spec:
  containers:
  - args:
    - /bin/sh
    - -c
    - echo hello;sleep 3600
    image: xxxxxxxxxxxx.dkr.ecr.us-east-2.amazonaws.com/simple-app:0.1.0
    imagePullPolicy: IfNotPresent
    name: test-fargate-app
    ports:
    - containerPort: 80
      protocol: TCP
    resources:
      limits:
        cpu: 250m
        memory: 512Mi
      requests:
        cpu: 250m
        memory: 512Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-bpxsq
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  priority: 2000001000
  priorityClassName: system-node-critical
  restartPolicy: Always
  schedulerName: fargate-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: default-token-bpxsq
    secret:
      defaultMode: 420
      secretName: default-token-bpxsq
status:
  nominatedNodeName: c62032c57f-3bf20f49a6d644568d59216432b8f221
  phase: Pending
  qosClass: Guaranteed
```

**Comment by nicholasgcoles** [comment]
Did you by any chance happen to change your `aws-auth` configmap?

I had an issue where I updated mine and there were behind the scenes mutable role bindings broken which no longer allowed for my pods to be scheduled

**Comment by khacminh** [comment]
It was the first thing that I checked, this is not mentioned on AWS document. Once maintaining my cluster,  I find out that EKS has to add a new map role to `aws-auth` after fargate profile is created to allow fargate pod to connect to my cluster.

```yaml
apiVersion: v1
data:
  mapRoles: |
    - groups:
      - system:bootstrappers
      - system:nodes
      rolearn: arn:aws:iam::xxxxxxxxxxxx:role/my-worker-node-role
      username: system:node:{{EC2PrivateDNSName}}
    - groups:
      - system:bootstrappers
      - system:nodes
      - system:node-proxier
      rolearn: arn:aws:iam::xxxxxxxxxxxx:role/eks-fargate-pods
      username: system:node:{{SessionName}}
```

**Comment by nicholasgcoles** [comment]
@khacminh this is true. I found that there was still something else going on in the background besides additional role mappings,

So if I ran `kubectl apply -f aws-auth.yaml`, and had the "correct" role mappings it still broke because of some immutable dependency (I had to open up an AWS support case for them to tell me this).

Obviously not ideal, but if you remove and then re-add your fargate profile does it work?

**Comment by khacminh** [comment]
@nicholasgcoles 
I did it before and try again today but the issue is till there.

**Comment by mikestef9** [comment]
Hi @khacminh we are tracking Fargate visibility in the console as part of #640. Have you opened a support case for the issue you are facing?

**Comment by khacminh** [comment]
Hi @mikestef9,

- How about creating fargate log streams to cloudwatch? Logs are always a life saver.
- It could be great if AWS could provide a detail documentation of fargate as the way Kubernetes describe its items.

---

About my problem, I found out that it the fargate node cannot resolve the DNS of STS in my region. That's why the fargate nodes cannot join my cluster. After checking my VPC, the VPC configuration wasn't updated for months and I created a STS endpoint Interface in the VPC before EKS fargate released. A few minute after deleting the STS endpoint interface in my VPC, the fargate node could join the cluster.

**Comment by tommy-couzens** [comment]
We are having the same issue with the STS endpoint.

When adding a STS VPC endpoint pods get stuck on pending and no fargate nodes can join the cluster.

**Comment by khacminh** [comment]
@tommy-couzens it took me days for troubleshooting because there is no fargate log

**Comment by tpmurthy** [comment]
> @tommy-couzens it took me days for troubleshooting because there is no fargate log

@khacminh How did you troubleshoot your issue? My pods are in 'Pending' status. I do not seem to have an STS Endpoint interface. Could you provide any tips to troubleshoot this issue?

**Comment by khacminh** [comment]
@tpmurthy 

I discover this while trying to create a new cluster.
I have an **ops** machine that used for my team to run the **kubectl** commands. Then I provided it enough Permissions to run **eksctl** command to create a new cluster with the same configurations (VPC, Subnets, IAM...). Then the **eksctl** threw out some errors that make it cannot create a new cluster. I traced the error to find out the problem.
